src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp

```diff
@@ -0,0 +1,519 @@
+};
+ATTRIBUTE_ALIGNED(32) static const uint64_t CONST_1p260[] = {
+    0x5030000000000000ULL,    // 0x1p+260
+};
+
+ATTRIBUTE_ALIGNED(32) static const uint64_t CONST_MAX[] = {
+    0x7FEFFFFFFFFFFFFFULL,    // Max
+};
+
+ATTRIBUTE_ALIGNED(32) static const uint64_t CONST_INF[] = {
+    0x7FF0000000000000ULL,    // Inf
+};
+
+ATTRIBUTE_ALIGNED(32) static const uint64_t CONST_e307[] = {
+    0x7FE0000000000000ULL
+};
+
+address StubGenerator::generate_libmFmod() {
+  StubCodeMark mark(this, "StubRoutines", "libmFmod");
+  address start = __ pc();
+  __ enter(); // required for proper stackwalking of RuntimeStub frame
+
+  if (VM_Version::supports_avx512vlbwdq()) {     // AVX512 version
+
+    // Source used to generate the AVX512 fmod assembly below:
+    //
+    // #include <ia32intrin.h>
+    // #include <emmintrin.h>
+    // #pragma float_control(precise, on)
+    //
+    // #define UINT32 unsigned int
+    // #define SINT32 int
+    // #define UINT64 unsigned __int64
+    // #define SINT64 __int64
+    //
+    // #define DP_FMA(a, b, c)    __fence(_mm_cvtsd_f64(_mm_fmadd_sd(_mm_set_sd(a), _mm_set_sd(b), _mm_set_sd(c))))
+    // #define DP_FMA_RN(a, b, c)    _mm_cvtsd_f64(_mm_fmadd_round_sd(_mm_set_sd(a), _mm_set_sd(b), _mm_set_sd(c), (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC)))
+    // #define DP_FMA_RZ(a, b, c) __fence(_mm_cvtsd_f64(_mm_fmadd_round_sd(_mm_set_sd(a), _mm_set_sd(b), _mm_set_sd(c), (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC))))
+    //
+    // #define DP_ROUND_RZ(a)   _mm_cvtsd_f64(_mm_roundscale_sd(_mm_setzero_pd(), _mm_set_sd(a), (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC)))
+    //
+    // #define DP_CONST(C)    _castu64_f64(0x##C##ull)
+    // #define DP_AND(X, Y)   _mm_cvtsd_f64(_mm_and_pd(_mm_set_sd(X), _mm_set_sd(Y)))
+    // #define DP_XOR(X, Y)   _mm_cvtsd_f64(_mm_xor_pd(_mm_set_sd(X), _mm_set_sd(Y)))
+    // #define DP_OR(X, Y)    _mm_cvtsd_f64(_mm_or_pd(_mm_set_sd(X), _mm_set_sd(Y)))
+    // #define DP_DIV_RZ(a, b) __fence(_mm_cvtsd_f64(_mm_div_round_sd(_mm_set_sd(a), _mm_set_sd(b), (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC))))
+    // #define DP_FNMA(a, b, c)    __fence(_mm_cvtsd_f64(_mm_fnmadd_sd(_mm_set_sd(a), _mm_set_sd(b), _mm_set_sd(c))))
+    // #define DP_FNMA_RZ(a, b, c) __fence(_mm_cvtsd_f64(_mm_fnmadd_round_sd(_mm_set_sd(a), _mm_set_sd(b), _mm_set_sd(c), (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC))))
+    //
+    // #define D2L(x)  _mm_castpd_si128(x)
+    // // transfer highest 32 bits (of low 64b) to GPR
+    // #define TRANSFER_HIGH_INT32(X)   _mm_extract_epi32(D2L(_mm_set_sd(X)), 1)
+    //
+    // double fmod(double x, double y)
+    // {
+    // double a, b, sgn_a, q, bs, bs2;
+    // unsigned eq;
+
+    Label L_5280, L_52a0, L_5256, L_5300, L_5320, L_52c0, L_52d0, L_5360, L_5380, L_53b0, L_5390;
+    Label L_53c0, L_52a6, L_53d0, L_exit;
+
+    __ movdqa(xmm2, xmm0);
+    //     // |x|, |y|
+    //     a = DP_AND(x, DP_CONST(7fffffffffffffff));
+    __ movq(xmm0, xmm0);
```

Comment:
I do not believe these are redundant, as the upper quadword of the register is cleared as a side-effect of the vmovq.  I do not believe the icx compiler would insert random redundant vmovq instructions at this optimization level.

Reply:
Subsequent uses of xmm0 operate on 128 bit vector and eventually it feed into DIVSD instruction operating on fist 64 bit data. Given that we are clearing upper 64 bit it may be issued to execution port and consume 1 cycle.